Zend/zend_compile.c

```diff
@@ -4239,6 +4239,12 @@ void zend_compile_throw(zend_ast *ast) /* {{{ */
 	znode expr_node;
 	zend_compile_expr(&expr_node, expr_ast);
 
+	if (CG(active_op_array)->fn_flags & ZEND_ACC_HAS_THROWS_TYPE) {
+		zend_op *opline = zend_emit_op(NULL, 
+			ZEND_VERIFY_THROW_TYPE, &expr_node, NULL);
+		zend_alloc_cache_slot(opline->op2.constant);
+	}
```

Comment:
This is only going to check exceptions thrown directly from the function, but not those that bubble up from a lower stack frame. You probably want to introduce this check in ZEND_HANDLE_EXCEPTION.